lib/SILOptimizer/Analysis/EscapeAnalysis.cpp

```diff
@@ -383,248 +409,362 @@ EscapeAnalysis::ConnectionGraph::getNode(ValueBase *V, bool createIfNeeded) {
+    if (node->pointsTo) {
+      assert(node->pointsTo->getMergeTarget() == newPointsTo);
+      // Since this node already had a pointsTo, it must reach a pointsTo
+      // edge. Stop traversing the defer-web here--this is complete becaused
+      // nodes are initialized one at a time, each time a new defer edge is
+      // created. If this were not complete, then the backward traversal below
+      // in Step 2 could reach uninitialized nodes not seen here in Step 1.
+      pointsToEdges.push_back(node);
+      return true;
     }
-  }
-  return true;
+    ++updateCount;
+    if (node->defersTo.empty()) {
+      // If this node is the end of a defer-edge path with no pointsTo
+      // edge. Create a "fake" pointsTo edge to maintain the graph invariant
+      // (this changes the structure of the graph but adding this edge has no
+      // effect on the process of merging nodes or creating new defer edges).
+      pointsToEdges.push_back(node);
+    }
+    updatedNodes.push(node);
+    if (!isSuccessor)
+      backwardReachable.insert(node);
+
+    return true;
+  };
+  // updatedNodes may grow during this loop.
+  unsigned nextUpdatedNodeIdx = 0;
+  for (; nextUpdatedNodeIdx < updatedNodes.size(); ++nextUpdatedNodeIdx)
+    updatedNodes[nextUpdatedNodeIdx]->visitDefers(visitDeferTarget);
+  // Reset this worklist so others can be used, but updateNode.nodeVector still
+  // holds all the nodes found by step 1.
+  updatedNodes.reset();
+
+  // Step 2: Update pointsTo fields by propagating backward from nodes that
+  // already have a pointsTo edge.
+  assert(nextUpdatedNodeIdx == updatedNodes.size());
+  --nextUpdatedNodeIdx;
+  bool processBackwardReachable = false;
+  do {
+    while (!pointsToEdges.empty()) {
+      CGNode *edgeNode = pointsToEdges.pop_back_val();
+      if (!edgeNode->pointsTo) {
+        edgeNode->setPointsToEdge(newPointsTo);
+        newPointsTo->mergeUsePoints(edgeNode);
+        assert(updateCount--);
+      }
+      // If edgeNode is already set to newPointsTo, it either was already
+      // up-to-date before calling initializePointsTo, or it was visited during
+      // a previous iteration of the backward traversal below. Rather than
+      // distinguish these cases, always retry backward traversal--it just won't
+      // revisit any edges in the later case.
+      backwardTraverse(edgeNode, [&](Predecessor pred) {
+        if (pred.getInt() != EdgeType::Defer)
+          return Traversal::Backtrack;
+
+        CGNode *predNode = pred.getPointer();
+        if (predNode->pointsTo) {
+          assert(predNode->pointsTo->getMergeTarget() == newPointsTo);
+          return Traversal::Backtrack;
+        }
+        predNode->pointsTo = newPointsTo;
+        newPointsTo->mergeUsePoints(predNode);
+        assert(updateCount--);
+        return Traversal::Follow;
+      });
+    }
+    // For all nodes visited in step 1, if any node was not backward-reachable
```

Comment:
This is for handling cycles which don't have a points-to edge, right? Like this was done in the old version in updatePointsTo inside ' if (isInitialSet) { }'
If yes, can you add a comment here?

Reply:
It's not really about cycles, which is one of the things that confused me about the old code. I still don't like this `initializePointsTo` code because it's still hard to figure out from a quick read. The concept is nice though:

Step 1: find all nodes in the uninitialized defer web and mark which ones already have edges.

Step 2: A direct application of the graph invariant "defer paths end in a pointsTo edge"

2a: walk backward from the edges checking that all the other nodes are reachable.

2b: add an edge to one of the remaining nodes and repeat

I was trying to be super efficient about this initially, then added extra intelligence later in (2b) to pick the best remaining node without changing the algorithmic complexity. But in hindsight it would be easier to understand the code if instead step (2b) just iterated over the remaining nodes again and directly identified the "leaves" in the graph.